RoomDeterminism/RoomDeterminism.lua

```diff
@@ -0,0 +1,321 @@
+      local fountain_room_name = BIOME_ABBREVIATION[biome] .. "_Reprieve01"
+      repeat
+        fountain_depth = rng:Random(config.FountainMinDepth[biome], config.FountainMaxDepth[biome])
+      until #(placed_rooms[fountain_depth] or {}) < ((placed_rooms[fountain_depth] or {NumExits = MAX_EXITS[biome]}).NumExits or MAX_EXITS[biome]) -- Previous room has enough exits for this room
+            and #(placed_rooms[fountain_depth + 1] or {}) <= RoomSetData[biome][fountain_room_name].NumExits -- This room can support the number of needed exits for next room
+      placed_rooms[fountain_depth] = placed_rooms[fountain_depth] or {} -- Create if nil
+      table.insert(placed_rooms[fountain_depth], fountain_room_name)
+    end
+
+    -- Roll Story
+    local has_story = RandomChance(config.StoryChance[biome])
+
+    -- Place Story
+    if has_story then
+      local story_room_name = BIOME_ABBREVIATION[biome] .. "_Story01"
+      repeat
+        story_depth = rng:Random(config.StoryMinDepth[biome], config.StoryMaxDepth[biome])
+      until #(placed_rooms[story_depth] or {}) < ((placed_rooms[story_depth] or {NumExits = MAX_EXITS[biome]}).NumExits or MAX_EXITS[biome]) -- Previous room has enough exits for this room
+            and #(placed_rooms[story_depth + 1] or {}) <= RoomSetData[biome][story_room_name].NumExits -- This room can support the number of needed exits for next room
+      placed_rooms[story_depth] = placed_rooms[story_depth] or {} -- Create if nil
+      table.insert(placed_rooms[story_depth], story_room_name)
+    end
+
+    -- For each depth, excluding the end shop, end boss and midbiome
+    for depth = BIOME_DEPTH[biome].Min, BIOME_DEPTH[biome].Max - 3 do
+      local minimum_exits = #(placed_rooms[depth + 1] or {1})
+      local eligible_room_data = GetEligibleCombatRoomData(biome, depth, minimum_exits)
+
+      -- Select room from eligible rooms
+      local selected_room_name = GetRandomKey(eligible_room_data)
+      RoomOverridesByDepth[depth] = RoomOverridesByDepth[depth] or {} -- Create if nil
+      RoomOverridesByDepth[depth].Room = selected_room_name
+      SelectedRoomsCache[selected_room_name] = true
+      DebugPrint({Text = depth .. ": Selected " .. selected_room_name})
+
+      -- Generate a table of options based on number of exits of selected room
+      local exit_options = {}
+      for i = 1, eligible_room_data[selected_room_name].NumExits do
+        table.insert(exit_options, i)
+      end
+
+      -- Assign exits to any placed rooms at the next depth
+      if #(placed_rooms[depth + 1] or {}) > 0 then
+        for _, placed_room in ipairs(placed_rooms[depth + 1]) do
+          local exit = RemoveRandomValue(exit_options)
+          RoomOverridesByDepth[depth]["Exit" .. exit] = placed_room
+          DebugPrint({Text = depth .. ": Placed " .. placed_room .. " at " .. "Exit" .. exit})
+        end
+      end
+    end
+  end
+end
+
+-- TODO: How to handle RequiredFalseRooms / Combat08A in tartarus?
```

Comment:
If you pass in the list of selected rooms at depth n-1, then we can disallow Combat08A if any of them are in the list.